src/nvim/regexp.c

```diff
@@ -7036,29 +7033,63 @@ void vim_regfree(regprog_T *prog)
     prog->engine->regfree(prog);
 }
 
-/*
- * Match a regexp against a string.
- * "rmp->regprog" is a compiled regexp as returned by vim_regcomp().
- * Uses curbuf for line count and 'iskeyword'.
- *
- * Return TRUE if there is a match, FALSE if not.
- */
-int 
-vim_regexec (
-    regmatch_T *rmp,
-    char_u *line,      /* string to match against */
-    colnr_T col            /* column to start looking for match */
-)
+static void report_re_switch(char_u *pat)
 {
-  return rmp->regprog->engine->regexec_nl(rmp, line, col, false);
+  if (p_verbose > 0) {
+    verbose_enter();
+    MSG_PUTS(_("Switching to backtracking RE engine for pattern: "));
+    MSG_PUTS(pat);
+    verbose_leave();
+  }
 }
 
-/*
- * Like vim_regexec(), but consider a "\n" in "line" to be a line break.
- */
+/// Match a regexp against a string.
+/// "rmp->regprog" is a compiled regexp as returned by vim_regcomp().
+/// Uses curbuf for line count and 'iskeyword'.
+/// When "nl" is TRUE consider a "\n" in "line" to be a line break.
+///
+/// Return TRUE if there is a match, FALSE if not.
+///
+/// @param rmp
+/// @param line string to match against
+/// @param col column to start looking for match
+/// @param nl
+static int vim_regexec_both(regmatch_T *rmp, char_u *line, colnr_T col, int nl)
+{
+  int result = rmp->regprog->engine->regexec_nl(rmp, line, col, nl);
+
+  // NFA engine aborted because it's very slow.
+  if (rmp->regprog->re_engine == AUTOMATIC_ENGINE
+      && result == NFA_TOO_EXPENSIVE) {
+    int save_p_re = p_re;
+    int re_flags = rmp->regprog->re_flags;
+    char_u *pat = vim_strsave(((nfa_regprog_T *)rmp->regprog)->pattern);
+
+    p_re = BACKTRACKING_ENGINE;
+    vim_regfree(rmp->regprog);
+    if (pat != NULL) {
```

Comment:
@splinterofchaos Might be a stupid question but then what happens when we run out of memory in `vim_strsave()` (presumably that's the situation where it would return `NULL`)?